### PR TITLE
network: on premise Service load balancers

### PR DESCRIPTION
**Pull request by russellb**
OpenShift does not currently support Services of type=LoadBalancer on
bare metal or other on premise infrastructure environments.  This
enhancement proposes a way forward, which is to adopt MetalLB.

The document explores some high level requirements, discusses some
alternatives considered, and maps out how we would get to work on
this through technical due diligence, upstream community engagement,
and careful planning of OpenShift integration with a new operator.

There is not currently a target for when this would be fully
supported, as this is a proposed enhancement on a direction to take.
More extensive technical due diligence, development, and testing will
help define the roadmap over time.

### CODE REVIEW

File: enhancements/network/on-prem-service-load-balancers.md [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+support](https://metallb.universe.tf/configuration/calico/).
+
+The layer 2 mode has the advantage of working in more environments.  We could
+also consider a MetalLB enhancement that makes it understand different L2
+domains and manage different IP address pools for each domain where SLBs may
+reside.
+
+### User Stories
+
+#### Story 1 - Easy Use with a Small Cluster
```
Comment (smarterclayton) [inline enhancements/network/on-prem-service-load-balancers.md:96]: We should define maybe Story 0 that is "As an application deployer on top of metal, a bare metal UPI cluster supports service load balancers exactly like a cloud service load balancer does, including in cluster access, out of cluster access, the possibility of DNS, the e2e tests working exactly like they do for AWS, etc etc etc".  I think it's important to define that the goal of this is to at minimum be consistent with cloud, then broaden into domain specific problems.

File: enhancements/network/on-prem-service-load-balancers.md [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+by the operator.
+
+We must also mitigate these risks through engagement and contributions to the
+upstream community to help make sure that changes made to the software and its
+configuration interfaces can be managed through an upgrade or downgrade
+process.
+
+### Version Skew Strategy
+
+MetalLB has two major components: a single `controller` pod and a `speaker` pod
```
Comment (derekwaynecarr) [inline enhancements/network/on-prem-service-load-balancers.md:222]: does the `speaker` run on every node or a subset?

File: enhancements/network/on-prem-service-load-balancers.md [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+title: on-prem-service-load-balancers
+authors:
+  - "@russellb"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-05-14
+last-updated: 2020-05-14
+status: implementable
```
Comment (derekwaynecarr) [inline enhancements/network/on-prem-service-load-balancers.md:11]: is this implementable yet or is it more proposed?  i would expect more implementation details if implementable, but that is probably separate enhancement tied to the operator.

File: enhancements/network/on-prem-service-load-balancers.md [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+    * Layer 2 (gratuitous ARP for IPv4, NDP for IPv6) - good for wide range of
+      environment compatibility, but limiting for larger clusters
+    * Layer 3 (BGP) - good for integration with networks for larger clusters
+* Suitable for large scale clusters (target up to 2000 nodes).
+* Single Node IPs (all traffic for a Service LB going through a single Node) is
+  acceptable for our out-of-the-box solution.  Support for a multi-Node based
+  solution would be nice, but is not required.
+* Must be compatible with at least the following cluster network types:
+  [OpenShift-SDN](https://github.com/openshift/sdn) and
+  [OVN-Kubernetes](https://github.com/openshift/ovn-kubernetes)
```
Comment (danwinship) [inline enhancements/network/on-prem-service-load-balancers.md:59]: use the upstream link. https://github.com/ovn-org/ovn-kubernetes
(we don't want a solution that depends on patches in our fork)

File: enhancements/network/on-prem-service-load-balancers.md [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+
+### Implementation Details/Notes/Constraints
+
+The first step of implementation is to invest in the upstream project.  We
+should have one or more engineers engage with the project to handle issues,
+review pull requests, and contribute bug fixes or enhancements.  As part of
+this process we should continue our technical due diligence with testing and
+reviewing code to increase our confidence in choosing this solution.  One area
+we could contribute immediately is with setting up upstream CI.  The project
+does not appear to run any CI today.
```
Comment (danwinship) [inline enhancements/network/on-prem-service-load-balancers.md:139]: Based on our experiences in ovn-kubernetes upstream, I think [kind](https://github.com/kubernetes-sigs/kind/) is a good basis for CI of community kubernetes-ecosystem projects. It's not nearly as fancy as what we do in kubernetes and openshift, but you can use it with the built-in free github CI support rather than needing someone to be paying for an AWS/GCP account, etc.

(It would even allow testing both L2 and L3 mode. It doesn't matter what protocols the actual underlying network supports if you're doing all of your testing in a virtual network built on top of it.)

File: enhancements/network/on-prem-service-load-balancers.md (hunk 1 above)
Comment (danwinship) [reply enhancements/network/on-prem-service-load-balancers.md:96]: We can't make _all_ bare metal UPI clusters work like cloud environments (eg, there's no plan for clusters where the nodes are on multiple L2 segments and you can't use BGP between them). But Russell could merge the "exactly like a cloud service" text into his existing stories.

Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:96]: I'll look at how to incorporate the "exactly like a cloud service" input into the text.

Regarding clusters on multiple L2 segments and no BGP available, I think that's an enhancement someone could implement.  I've seen it requested on issues, and someone wrote a PR to do it (https://github.com/metallb/metallb/pull/502), though it needs some further work.  I'd like to see something done for this, because it seems like a common enough use case.

File: enhancements/network/on-prem-service-load-balancers.md (hunk 2 above)
Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:222]: Good question!

I believe "it depends".  In layer 2 mode, it really does need to run on every Node.

In BGP mode, there's more flexibility.  The behavior depends on the `ExternalTrafficPolicy` type on the `Service`.  
- If the type is `cluster`, then `speaker` can run on any subset of Nodes that you want peering with BGP routers.  All of those Nodes will advertise that they can be used to reach all of of the Service load balancer IPs.
- If the type is `local`, then `speaker` must run on every node where an `Endpoint` may exist locally.  Each node will only advertise that load balancer IPs are reachable that can map to a local Endpoint.  Put another way, `speaker` must run on every node where workloads behind a service load balancer may run.

I'll incorporate this information into the enhancement.

File: enhancements/network/on-prem-service-load-balancers.md [hunk 6]
```diff
@@ -0,0 +1,370 @@
+---
+title: on-prem-service-load-balancers
+authors:
+  - "@russellb"
+reviewers:
+  - TBD
+approvers:
+  - TBD
```
[the remark this answers is not in the document]
Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:8]: I don't know.  I wasn't sure, left it blank, and forgot to come back to it.

some thoughts ...

- someone from architecture team (you if interested, @derekwaynecarr ?  or @smarterclayton indicated he has a lot of interest in this topic)
- networking group lead (@knobunc )
- someone from team likely to own this (kni networking @celebdor 's team)
- lead from networking edge team, since they own service load balancers more generally

File: enhancements/network/on-prem-service-load-balancers.md (hunk 3 above)
Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:11]: I guess `proposed` is more appropriate since it doesn't get into the more specific details around the operator.  This is more about kicking off the direction, starting to engage upstream, and starting on operator design (but not specifying the design itself)

File: enhancements/network/on-prem-service-load-balancers.md (hunk 4 above)
Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:59]: will do, thanks

File: enhancements/network/on-prem-service-load-balancers.md (hunk 5 above)
Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:139]: Awesome, thanks for the feedback!  I'll incorporate this into the document.

File: enhancements/network/on-prem-service-load-balancers.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:8]: since @danwinship already commented, maybe you'd have interest in approving this as a representative from the networking group?

[the remark this answers is not in the document]
Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:8]: ... and it occurs to me there's no guidance on how to fill out reviewers/approvers in the template, so i'll follow up on that separatement

**Comment by russellb** [comment]
Note: I will squash commits once review has completed.

File: enhancements/network/on-prem-service-load-balancers.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (danwinship) [reply enhancements/network/on-prem-service-load-balancers.md:8]: sure

File: enhancements/network/on-prem-service-load-balancers.md (hunk 1 above)
Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:104]: I've captured this, thanks

File: enhancements/network/on-prem-service-load-balancers.md [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+We do not currently support [Services of
```
Comment (squeed) [inline enhancements/network/on-prem-service-load-balancers.md:26]: Just FYI, we have a funny feature around this: "IngressIPs", where we automatically assign ExternalIPs to services of type LoadBalancer if configured. See https://github.com/openshift/api/blob/master/config/v1/types_network.go#L99

File: enhancements/network/on-prem-service-load-balancers.md [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+network segments, I would like to configure one or more ranges of IP addresses
+from my network that the cluster is free to use for Service Load Balancers.  I
+would like my cluster to peer with my BGP infrastructure to advertise the
+current location of IP addresses allocated to Service Load Balancers.
+
+MetalLB can do this today.
+
+#### Story 3 - Larger Clusters without BGP
+
+As an administrator of a larger cluster with Nodes that reside on multiple L2
```
Comment (squeed) [inline enhancements/network/on-prem-service-load-balancers.md:126]: I know that the API service is announced using keepalived + HAProxy. I don't know how manual this solution is, however.

Comment (squeed) [reply enhancements/network/on-prem-service-load-balancers.md:126]: And you mentioned that below :)

File: enhancements/network/on-prem-service-load-balancers.md [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+Despite the project's young age, someone has already [integrated it with
+OpenShift 3.11](https://github.com/megian/openshift-kube-vip-ansible).
+
+Some of the key downsides to this option:
+* Depends on yet-another-raft-implementation --
+  https://github.com/hashicorp/raft
+* New, developed as a hobby project by one person, likely PoC level maturity
+* Lacking any layer 3 based address advertisement options
+
+### OVN-Kubernetes Native Solution
```
Comment (squeed) [inline enhancements/network/on-prem-service-load-balancers.md:366]: This is an interesting proposal, and I agree it's probably not the right solution right now, but it's definitely interesting as  a motivating user story for the future. FYI @dcbw.

File: enhancements/network/on-prem-service-load-balancers.md (hunk 8 above)
Comment (celebdor) [reply enhancements/network/on-prem-service-load-balancers.md:126]: the only manual step is specifying two IP addresses at installation time.

File: enhancements/network/on-prem-service-load-balancers.md [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+
+Keepalived only supports L2 advertisement of IP address location (ARP / NDP).
+To support larger scale clusters, we must do one of the following:
+
+* Require all SLBs to be hosted on Nodes within a single L2 domain within the
+  cluster.
+* Make our SLB controller smart enough to understand different IP address
+  pools, their associated L2 domains, and which Nodes are on which L2 domain.
+* Extend keepalived (either directly or via some integration) to support an L3
+  based address location advertisement (likely BGP).
```
Comment (celebdor) [inline enhancements/network/on-prem-service-load-balancers.md:311]: It is more likely that we'd use bird or quagga when not on GARP mode.

File: enhancements/network/on-prem-service-load-balancers.md [hunk 11]
```diff
@@ -0,0 +1,405 @@
...
+* Management of one or more pools of IP addresses to be allocated for SLBs
+* High Availability (HA) management of these IP addresses once allocated.  We
+  must be able to fail over addresses in less than 5 seconds.
+* Advertisement of the location of an IP address once assigned to a Node or
+  moved to a new Node.  Possible approaches include:
+    * Layer 2 (gratuitous ARP for IPv4, NDP for IPv6) - good for wide range of
+      environment compatibility, but limiting for larger clusters
+    * Layer 3 (BGP) - good for integration with networks for larger clusters
+* Suitable for large scale clusters (target up to 2000 nodes).
+* Single Node IPs (all traffic for a Service LB going through a single Node) is
```
Comment (abhinavdahiya) [inline enhancements/network/on-prem-service-load-balancers.md:54]: One of the mandate for upgrades is that there must be no disruption of services like APIserver or Ingress. Single Node IPs will probably fail to provide that right? the node will be drained and rebooted and then there will probably disruption of traffic while the IP moves to another node. https://github.com/openshift/enhancements/pull/356/files#diff-4efb0274cd0d2bdbac261b1c4f44cbacR46 points to 5 secs of fail over.

Is it possible to include no disruption of traffic when upgrading a goal? this would help target a specific case and not generic solution.

File: enhancements/network/on-prem-service-load-balancers.md [hunk 12]
```diff
@@ -0,0 +1,405 @@
...
+capabilities.
+
+### Goals
+
+A SLB solution must provide these high level features:
+
+* Management of one or more pools of IP addresses to be allocated for SLBs
+* High Availability (HA) management of these IP addresses once allocated.  We
+  must be able to fail over addresses in less than 5 seconds.
+* Advertisement of the location of an IP address once assigned to a Node or
```
Comment (abhinavdahiya) [inline enhancements/network/on-prem-service-load-balancers.md:48]: this goal seems to be tied a little bit more tied to the actual impl of metallb. do we really care what advertisement is used? Maybe we reword it to match specific environments that must be able to use our solution?

File: enhancements/network/on-prem-service-load-balancers.md (hunk 11 above)
Comment (cybertron) [reply enhancements/network/on-prem-service-load-balancers.md:54]: When the apiserver goes down gracefully it notifies loadbalancers through the /readyz endpoint before it actually stops responding. We would probably want something similar for nodes being drained, a way to determine that it is happening so we can move the IP before it actually reboots and stops responding. That should avoid the failover delay because we'd be moving the IP from one still-functional node to another.

File: enhancements/network/on-prem-service-load-balancers.md (hunk 7 above)
Comment (aojea) [reply enhancements/network/on-prem-service-load-balancers.md:26]: external IP can be set regardless of service type. For LoadBalancer services you have to set the ingress field in status.
```
status:
  loadBalancer:
    ingress:
    - ip: 203.0.113.1
```

File: enhancements/network/on-prem-service-load-balancers.md [hunk 13]
```diff
@@ -0,0 +1,370 @@
...
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+We do not currently support [Services of
+type=LoadBalancer](https://kubernetes.io/docs/concepts/services-networking/#loadbalancer)
+(Service Load Balancers, or SLBs) for OpenShift in bare metal environments.
+While bare metal clusters are of primary interest, we hope to find a solution
+that would apply to other on-premise environments that don’t have native load
+balancer capabilities available (a cluster on VMware, RHV, or OpenStack, for
```
[the remark this answers is not in the document]
Comment (aojea) [reply enhancements/network/on-prem-service-load-balancers.md:31]: don't know if this is redundant, but there is one using the openstack-cloud-controller-manager for Octavia  https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/using-octavia-ingress-controller.md#why-octavia-ingress-controller

File: enhancements/network/on-prem-service-load-balancers.md [hunk 14]
```diff
@@ -0,0 +1,405 @@
...
+
+## Implementation History
+
+* (May, 2020) - Technical due diligence and upstream engagement beginning
+
+## Drawbacks
+
+TBD
+
+## Alternatives
```
Comment (aojea) [inline enhancements/network/on-prem-service-load-balancers.md:279]: I went down this path some months ago and let me share my findings, excuse me if I say something wrong because I don't have a good knowledge of Openshift yet.
I've come to the conclusion that the solution for bare metal should be based in a cloud-controller-manager, implementing the cloudprovider.Interface https://github.com/kubernetes/cloud-provider/blob/master/plugins.go and reusing the common logic there. Then, you can use whatever implementation for the load balancer, a good example of this approach is the `Packet` cloud controller manager, that uses MetalLB under the hood https://github.com/packethost/packet-ccm

It's a very neat approach IMHO with the benefit that you have a clear interface already defined upstream

File: enhancements/network/on-prem-service-load-balancers.md [hunk 15]
```diff
@@ -0,0 +1,405 @@
...
+[kind](https://github.com/kubernetes-sigs/kind/) is a good basis for CI of
+community kubernetes-ecosystem projects. It is usable with the built-in free
+github CI support rather than needing someone to be paying for test
+infrastructure elsewhere. It would even allow testing both L2 and L3 mode. It
+doesn't matter what protocols the actual underlying network supports if you're
+doing all of your testing in a virtual network built on top of it.
+
+#### Operator
+
+We must also create an operator for MetalLB.  We should develop an operator
```
Comment (aojea) [inline enhancements/network/on-prem-service-load-balancers.md:161]: See my comment https://github.com/openshift/enhancements/pull/356/files#r434739926 and the `Packet` controller https://github.com/packethost/packet-ccm

File: enhancements/network/on-prem-service-load-balancers.md (hunk 7 above)
Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:26]: Interesting, I didn't know about this feature.  I will incorporate this into the doc.  I imagine this was done before enhancements?  Any other docs I can refer to?

As you implied, I suppose the benefit of this proposal would be a way to make the IPs accessible automatically, either via ARP/NDP or BGP, so it still seems worth pursuing to me.  Would you agree?

File: enhancements/network/on-prem-service-load-balancers.md (hunk 8 above)
Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:126]: Yeah, I think this is discussed in the Alternatives section

File: enhancements/network/on-prem-service-load-balancers.md (hunk 9 above)
Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:366]: Thanks for that insight, @aojea.  At least with ovn-kubernetes there's also a central controller component that could take on the LoadBalancer management, so it wouldn't have to go into the CNI plugin directly.

File: enhancements/network/on-prem-service-load-balancers.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:31]: Yeah, I'll clarify this to say that it's only applicable to OpenStack in the case that Octavia is not available.  Octavia is not a hard requirement today, so I suspect we would want to keep it optional.

File: enhancements/network/on-prem-service-load-balancers.md [hunk 16]
```diff
@@ -0,0 +1,405 @@
...
+The first step of implementation is to invest in the upstream project.  We
+should have one or more engineers engage with the project to handle issues,
+review pull requests, and contribute bug fixes or enhancements.  As part of
+this process we should continue our technical due diligence with testing and
+reviewing code to increase our confidence in choosing this solution.
+
+One area we could contribute immediately is with setting up upstream CI.  The
+project does not appear to run any CI today.
+
+[kind](https://github.com/kubernetes-sigs/kind/) is a good basis for CI of
```
[the remark this answers is not in the document]
Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:152]: Thanks for the background!  Indeed, there seems to be an active transition of leadership from the original author to a new group that intends to carry it forward.

File: enhancements/network/on-prem-service-load-balancers.md (hunk 14 above)
Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:279]: @aojea great feedback, thank you!  I will explore this in more detail.

File: enhancements/network/on-prem-service-load-balancers.md (hunk 7 above)
Comment (jboxman) [reply enhancements/network/on-prem-service-load-balancers.md:26]: @russellb, this is [more about ExternalIP on 4.x](https://github.com/openshift/openshift-docs/pull/21388), but I haven't had time to complete it.

File: enhancements/network/on-prem-service-load-balancers.md (hunk 9 above)
Comment (aojea) [reply enhancements/network/on-prem-service-load-balancers.md:366]: > ovn-kubernetes there's also a central controller component that could take on the LoadBalancer management,

my point is that Load Balancers have different lifecycles and features, the release process and development is more complicated if you tie it to the CNI controller ... we made that mistake with Midonet too :sweat_smile:  @celebdor may remember those funny days :wink: and I think that Openstack did kind of the same with the neutron-lbaas project -> Octavia, but maybe wrong here and the reasons were others.
Let's assume you receive a new feature for the load balancer or a bug fix, it forces you to cut a new release for the ovn-kubernetes ... same happen once you want to roll back a new version or with an upgrade. You'll have different users, roadmaps and problems in the same code base :shrug:

File: enhancements/network/on-prem-service-load-balancers.md (hunk 14 above)
Comment (danwinship) [reply enhancements/network/on-prem-service-load-balancers.md:279]: > I've come to the conclusion that the solution for bare metal should be based in a cloud-controller-manager, implementing the cloudprovider.Interface

But you can only have one cloud provider, can't you? The doc says they ideally want something that will work with OpenStack and VMware too, where presumably you'd want to keep using the native (load-balancer-less) cloud provider.

Comment (aojea) [reply enhancements/network/on-prem-service-load-balancers.md:279]: ok, my fault, I should've have explained myself better, because I'm considering that we have a control-plane piece (cloud-controller-manager) and a data-plane, envoy, metallb uses bgp or keepalived like load balancing, ..ovn , 
I think makes sense to have a different cloud-controller-manager per environment, i.e aws, gcp, azure, openstck, vsphere, ..., actually this is how it is working upstream.
However, the dataplane mechanism may be the same for all ... if the load balancer is going to be as simple as metallb it makes sense to use OVN, just bear in mind that if you expect demand for real load balancing features (SSL/TLS, grpc, affinity, health checking, ... ) then envoy is the king there and we start to move to the service mesh area ... 
It really depends on what you mean by load balancer: keepalived or envoy, for me load balancer is the later

File: enhancements/network/on-prem-service-load-balancers.md [hunk 17]
```diff
@@ -0,0 +1,405 @@
...
+example).
+
+## Motivation
+
+Service Load Balancers are a common way to expose applications on a cluster. We
+highly value clusters using on premise infrastructure, but do not support this
+feature in that context.  We aim to fill this gap with an optional range of
+capabilities.
+
+### Goals
```
Comment (yboaron) [inline enhancements/network/on-prem-service-load-balancers.md:41]: Do we need to define a goal for Network policy support?

File: enhancements/network/on-prem-service-load-balancers.md [hunk 18]
```diff
@@ -0,0 +1,405 @@
...
+
+#### Operator
+
+We must also create an operator for MetalLB.  We should develop an operator
+that is generally useful to the MetalLB community.  We should also have an
+OpenShift version of this operator for our use.
+
+It is assumed that the MetalLB operator would be managed by OLM as an optional
+additional component to be installed on on-premise clusters.  However, in the
+[ROADMAP.md
```
Comment (yboaron) [inline enhancements/network/on-prem-service-load-balancers.md:167]: Unlike LB provided by cloud-provider (e.g: Openstack Octavia), the logging/debugging capabilities of MetalLB are very limited. so, moving critical components such as api and ingress to rely on MetalLb might be an issue in that aspect.

File: enhancements/network/on-prem-service-load-balancers.md (hunk 7 above)
Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md]: I've added some notes about this existing functionality and clarified that we lack today is full automation for these services and not that they are completely unusable.  This enhancement adds built-in automation for making these IPs reachable in the right places.

File: enhancements/network/on-prem-service-load-balancers.md (hunk 11 above)
Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:65]: There's a new upstream issue tracking the possibility of adding a more graceful failover procedure: https://github.com/metallb/metallb/issues/621

I'll also look at updating the requirements to reflect that we want graceful no-downtime failover for planned upgrades.  I think that will already be fine for the layer 3 (BGP) mode, but would take some more work for the layer 2 mode.

File: enhancements/network/on-prem-service-load-balancers.md (hunk 12 above)
Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:59]: I'll reword this a bit.  Thanks for the feedback.

File: enhancements/network/on-prem-service-load-balancers.md (hunk 14 above)
Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:291]: I've added the CCM integration as an alternative integration path, but not specified what the conclusion is.  I assume that would come in a future update to this enhancement which proposes more detailed integration plans once further investigation is complete.

File: enhancements/network/on-prem-service-load-balancers.md (hunk 17 above)
Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:52]: I don't think MetalLB would be involved in NetworkPolicy.  MetalLB handles IPAM and ensuring traffic gets to the right node for a given load balancer IP.  Once the traffic gets there, it's up to the network solution in use to get the traffic to the right place.  That'd be kube-proxy or whatever its equivalent is in a given setup.

File: enhancements/network/on-prem-service-load-balancers.md (hunk 18 above)
Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:182]: I've added some notes on debugging as a risk

**Comment by russellb** [comment]
I believe I have addressed all open feedback.  The enhancement does not clarify a detailed plan for OpenShift integration, but it does highlight the options that must be explored in the next phase of this work.  The outcome I hope to reach here is consensus on whether MetalLB is a good direction to pursue.

The next step would be two things in parallel:

1) upstream investment
2) investigation for a detailed openshift integration plan, to be proposed as an update to this enhancement.  At that point, we could call this `implementable` instead of just `proposed`

As I noted earlier, I will squash this down to one commit before merging.

File: enhancements/network/on-prem-service-load-balancers.md (hunk 17 above)
Comment (yboaron) [reply enhancements/network/on-prem-service-load-balancers.md:52]: My bad, I wasn't clear. 
IIUC, MetalLB is responsible for: 
   1. IPAM & ensuring traffic gets to the right node for a given load balancer IP
   2. Adds some IPtable rules to manipulate the packet (change the external-IP) 

So, my concern was if the plugin responsible for NetworkPolicy should be aware to step '2' to be able to enforce correctly  the NetworkPolicy

Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:52]: AFAIK it's only doing point 1 and not point 2.  It's up to something else to make sure the right thing happens to the traffic once it hits the host with that external IP as a destination.

Comment (yboaron) [reply enhancements/network/on-prem-service-load-balancers.md:52]: You are right -  MetalLB doesn't create any IPtables rules.
I'll create a LB service in my env and investigate the IPtables rules.
Thanks

Comment (danwinship) [reply enhancements/network/on-prem-service-load-balancers.md:52]: Note that [the NetworkPolicy docs](https://kubernetes.io/docs/concepts/services-networking/network-policies/#behavior-of-to-and-from-selectors) explicitly allow basically any behavior on cluster ingress packets:

> Cluster ingress and egress mechanisms often require rewriting the source or destination IP of packets. In cases where this happens, it is not defined whether this happens before or after NetworkPolicy processing, and the behavior may be different for different combinations of network plugin, cloud provider, Service implementation, etc.
> 
> In the case of ingress, this means that in some cases you may be able to filter incoming packets based on the actual original source IP, while in other cases, the "source IP" that the NetworkPolicy acts on may be the IP of a LoadBalancer or of the Pod's node, etc.

File: enhancements/network/on-prem-service-load-balancers.md (hunk 14 above)
Comment (cybertron) [reply enhancements/network/on-prem-service-load-balancers.md:323]: (I realize this is marked resolved, but I wanted to document the discussions we've had offline)

One possible issue with using a CCM is that in metal3 we don't have a cloud API to talk to. Using a CCM would likely require work with the metal3 community to figure out how/if that would work.

**Comment by cybertron** [comment]
Just wanted to note that I have a PoC operator for deploying metallb. I did a quick demo of it here: https://www.youtube.com/watch?v=WgOZno0D7nw

**Comment by russellb** [comment]
> Just wanted to note that I have a PoC operator for deploying metallb. I did a quick demo of it here: https://www.youtube.com/watch?v=WgOZno0D7nw

Thanks.  I've added this to the doc, squashed all the commits, and rebased on master.

File: enhancements/network/on-prem-service-load-balancers.md [hunk 19]
```diff
@@ -0,0 +1,460 @@
...
+reachable.  This method would remain an option for any clusters where the
+administrator would like to manage routing for external IPs in a completely
+custom manner.  For more information on the existing `IngressIPs` feature:
+
+* https://github.com/openshift/api/blob/master/config/v1/types_network.go#L99
+* https://github.com/openshift/openshift-docs/pull/21388
+
+### Goals
+
+A SLB solution must provide these high level features:
```
Comment (markmc) [inline enhancements/network/on-prem-service-load-balancers.md:54]: Stupid question time!

This list of features does not mention ... load balancing? i.e. distributing incoming connections over a set of available endpoints?

I assume this means the actual load balancing is done by the service proxy provided by openshift-sdn or ovn-kubernetes. And the frontend IP for that load balancer is the clusterIP field. And MetaLB is all about allocating a "public" IP for a service and advertising that IP.

First point of confusion - if a connection for the MetalLB-allocated IP arrives on a node, how does it end up being accepted by a pod?

And then back to ... load-balancing - with BGP, is it just that multiple nodes advertise themselves as listening on the MetalLB IP, and the load-balancing across those multiple listeners depends on the router configuration. And with the L2 mode, only one of the listening nodes advertises itself?

I dunno, I think I'd appreciate some brief discussion of ... load-balancing? (kthxbi!)

Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:54]: > Stupid question time!

Not stupid at all.  This is by far the #1 point of confusion I've seen in the MetalLB community so far.  The docs really need to be improved to help clarify what MetalLB does and does not do, which funny enough does not include load balancing.

> This list of features does not mention ... load balancing? i.e. distributing incoming connections over a set of available endpoints?
> 
> I assume this means the actual load balancing is done by the service proxy provided by openshift-sdn or ovn-kubernetes. And the frontend IP for that load balancer is the clusterIP field. And MetaLB is all about allocating a "public" IP for a service and advertising that IP.

You got it.

> 
> First point of confusion - if a connection for the MetalLB-allocated IP arrives on a node, how does it end up being accepted by a pod?

Apparently service proxy implementations must already be ready to accept traffic to a Node with a destination IP address set to the external IP address for a LoadBalancer Service.  All MetalLB has to do is implement some environment-appropriate method of getting the traffic to a Node.

> 
> And then back to ... load-balancing - with BGP, is it just that multiple nodes advertise themselves as listening on the MetalLB IP, and the load-balancing across those multiple listeners depends on the router configuration. 

That is correct!  This is where "ECMP" comes in and we get real load balancing across Nodes for a single LoadBalancer.  You are correct that this relies on configuration of your router(s).

There's another detail to this ... there is a field called `externalTrafficPolicy` that can be `cluster` or `local`.   https://kubernetes.io/docs/tasks/access-application-cluster/create-external-load-balancer/#preserving-the-client-source-ip

* `local` -- In this mode, the pod expects to receive traffic with the original source IP address still intact.  To achieve this, the traffic must go directly to a Node where one of the Endpoints for that Service is running.  Only those Nodes advertise the Service IP in this case.

* `cluster` -- In this mode, traffic may arrive on any Node and will be redirected to another Node if necessary to reach a Service Endpoint.  The source IP address will be changed to the Node's IP to ensure traffic returns via the same path it arrived on.  The Service IP is advertised for *all* Nodes for these Services.

> And with the L2 mode, only one of the listening nodes advertises itself?

That is correct.  There's an algorithm used inside MetalLB that ensures that all Nodes agree which one it should be.
 
> I dunno, I think I'd appreciate some brief discussion of ... load-balancing? (kthxbi!)

ha, OK.  Yes, I'll take a fresh look at this doc and add some more information about it.  I've learned more since I first wrote it.

Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md:67]: I pushed another commit with some additional information on these topics.  I hope it helps.

Comment (markmc) [reply enhancements/network/on-prem-service-load-balancers.md:67]: Beautiful, that's super helpful!

File: enhancements/network/on-prem-service-load-balancers.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (russellb) [reply enhancements/network/on-prem-service-load-balancers.md]: I've updated this to include names for reviewers and approvers.  I've listed everyone who has commented so far as a reviewer.  I listed @knobunc , @danwinship , @danehans as approvers as representatives from different aspects of openshift networking.

I think this is fine to merge now if there are no new concerns.  Note that it's still marked "proposed", as I'd expect another enhancement to be filed to refine the operator plans.

**Comment by russellb** [comment]
@knobunc @danwinship @danehans The conversation on this one seems to have settled.  Note that it's in a "proposed" state.  I propose that we merge this and expect a follow-up later to address the remaining open questions around design details once we're ready to spend more time on this.
